File: ScriptsToProcess/NEW BATCH TO PROCESS/SegDisp.py
import Rhino.Geometry as rg
from Grasshopper import DataTree
from Grasshopper.Kernel.Data import GH_Path
import time

# --- Component Metadata ---
ghenv.Component.Name = "Curve Segment Dispatcher"
ghenv.Component.NickName = "SegDisp"
ghenv.Component.Description = "Explodes curves into Lines and Arcs, extracting Radii, Centers, and visual Dimensions."

# Start the timer
start_time = time.time()

# Initialize DataTrees
tree_lines = DataTree[object]()
tree_arcs = DataTree[object]()
tree_radii = DataTree[object]()
tree_centers = DataTree[object]()
tree_dims = DataTree[object]()

def process_geometry(crv, r_offset, show_dims, decimals, path):
    # Explode curve into segments
    segments = crv.DuplicateSegments()
    if not segments or len(segments) == 0:
        segments = [crv]

    for seg in segments:
        if seg is None: continue
        
        # 1. Handle Linear Segments
        if seg.IsLinear():
            tree_lines.Add(seg, path)
            
        # 2. Handle Arcs/Curves
        else:
            success, arc_prim = seg.TryGetArc(0.001)
            if success:
                tree_arcs.Add(seg, path)
                tree_radii.Add(arc_prim.Radius, path)
                tree_centers.Add(arc_prim.Center, path)
                
                # 3. Create Visual Dimensions if toggled ON
                if show_dims:
                    mid_angle = arc_prim.AngleDomain.Mid
                    pt_on_arc = arc_prim.PointAt(mid_angle)
                    
                    dir_vec = pt_on_arc - arc_prim.Center
                    dir_vec.Unitize()
                    
                    offset_pt = pt_on_arc + (dir_vec * r_offset)
                    
                    # Create leader line
                    leader_line = rg.Line(pt_on_arc, offset_pt).ToNurbsCurve()
                    
                    # Create Text Dot with Dynamic Decimals
                    fmt = "R {:." + str(decimals) + "f}"
                    label = fmt.format(arc_prim.Radius)
                    
                    text_dot = rg.TextDot(label, offset_pt)
                    
                    tree_dims.Add(leader_line, path)
                    tree_dims.Add(text_dot, path)

# --- Main Execution ---
# Sanitize Inputs
val_offset = offset if 'offset' in globals() and offset is not None else 10.0
val_toggle = dim_toggle if 'dim_toggle' in globals() and dim_toggle is not None else True
val_decimals = int(num_dec) if 'num_dec' in globals() and num_dec is not None else 2

if 'curves' in globals() and curves is not None:
    for i in range(curves.BranchCount):
        branch = curves.Branch(i)
        path = curves.Paths[i]
        for crv in branch:
            if crv: 
                process_geometry(crv, val_offset, val_toggle, val_decimals, path)

# Stop the timer and calculate duration
elapsed_ms = (time.time() - start_time) * 1000.0

# Update Component Message
status_str = "DIMS: ON" if val_toggle else "DIMS: OFF"
total_count = tree_lines.DataCount + tree_arcs.DataCount
# Line 1: Toggle & Decimal setting | Line 2: Stats & Time
ghenv.Component.Message = "{} ({} dec)\n{} Segs | {:.2f} ms".format(status_str, val_decimals, total_count, elapsed_ms)

# Assign to outputs
lines = tree_lines
arcs = tree_arcs
radii = tree_radii
centers = tree_centers
dims = tree_dims

File: Components/SegDisp.cs
using System;
using System.Linq;
using Grasshopper;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;
using Rhino.Geometry;

namespace Enzyme.Components
{
    public class SegDisp : GH_Component
    {
        public SegDisp()
          : base("Curve Segment Dispatcher", "SegDisp",
              "Explodes curves into Lines and Arcs, extracting Radii, Centers, and visual Dimensions.",
              "Enzyme", "Utilities")
        {
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddCurveParameter("curves", "curves", "Curves to process", GH_ParamAccess.tree);
            pManager.AddNumberParameter("offset", "offset", "Dimension offset", GH_ParamAccess.item, 10.0);
            pManager.AddBooleanParameter("dim_toggle", "dim_toggle", "Show dimensions", GH_ParamAccess.item, true);
            pManager.AddIntegerParameter("num_dec", "num_dec", "Number of decimals", GH_ParamAccess.item, 2);
            
            pManager[1].Optional = true;
            pManager[2].Optional = true;
            pManager[3].Optional = true;
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddCurveParameter("lines", "lines", "Linear segments", GH_ParamAccess.tree);
            pManager.AddCurveParameter("arcs", "arcs", "Arc segments", GH_ParamAccess.tree);
            pManager.AddNumberParameter("radii", "radii", "Arc radii", GH_ParamAccess.tree);
            pManager.AddPointParameter("centers", "centers", "Arc centers", GH_ParamAccess.tree);
            pManager.AddGenericParameter("dims", "dims", "Visual dimensions", GH_ParamAccess.tree);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            var watch = System.Diagnostics.Stopwatch.StartNew();

            if (!DA.GetDataTree(0, out GH_Structure<GH_Curve> curvesTree)) return;

            double offset = 10.0;
            DA.GetData(1, ref offset);

            bool dimToggle = true;
            DA.GetData(2, ref dimToggle);

            int numDec = 2;
            DA.GetData(3, ref numDec);

            var treeLines = new GH_Structure<GH_Curve>();
            var treeArcs = new GH_Structure<GH_Curve>();
            var treeRadii = new GH_Structure<GH_Number>();
            var treeCenters = new GH_Structure<GH_Point>();
            var treeDims = new GH_Structure<IGH_Goo>();

            foreach (var path in curvesTree.Paths)
            {
                var branch = curvesTree.get_Branch(path);
                
                // Cast branch elements explicitly as instructed
                foreach (GH_Curve ghCurve in branch.Cast<GH_Curve>())
                {
                    if (ghCurve == null || ghCurve.Value == null) continue;
                    
                    Curve crv = ghCurve.Value;
                    Curve[] segments = crv.DuplicateSegments();
                    if (segments == null || segments.Length == 0)
                    {
                        segments = new Curve[] { crv };
                    }

                    foreach (Curve seg in segments)
                    {
                        if (seg == null) continue;

                        if (seg.IsLinear())
                        {
                            treeLines.Append(new GH_Curve(seg), path);
                        }
                        else
                        {
                            if (seg.TryGetArc(out Arc arcPrim, 0.001))
                            {
                                treeArcs.Append(new GH_Curve(seg), path);
                                treeRadii.Append(new GH_Number(arcPrim.Radius), path);
                                treeCenters.Append(new GH_Point(arcPrim.Center), path);

                                if (dimToggle)
                                {
                                    double midAngle = arcPrim.AngleDomain.Mid;
                                    Point3d ptOnArc = arcPrim.PointAt(midAngle);

                                    Vector3d dirVec = ptOnArc - arcPrim.Center;
                                    dirVec.Unitize();

                                    Point3d offsetPt = ptOnArc + (dirVec * offset);

                                    Curve leaderLine = new Line(ptOnArc, offsetPt).ToNurbsCurve();
                                    
                                    string label = "R " + arcPrim.Radius.ToString("F" + numDec, System.Globalization.CultureInfo.InvariantCulture);
                                    TextDot textDot = new TextDot(label, offsetPt);

                                    treeDims.Append(new GH_Curve(leaderLine), path);
                                    treeDims.Append(new GH_ObjectWrapper(textDot), path);
                                }
                            }
                        }
                    }
                }
            }

            watch.Stop();
            double elapsedMs = watch.Elapsed.TotalMilliseconds;

            string statusStr = dimToggle ? "DIMS: ON" : "DIMS: OFF";
            int totalCount = treeLines.DataCount + treeArcs.DataCount;
            
            this.Message = string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0} ({1} dec)\n{2} Segs | {3:F2} ms", statusStr, numDec, totalCount, elapsedMs);

            DA.SetDataTree(0, treeLines);
            DA.SetDataTree(1, treeArcs);
            DA.SetDataTree(2, treeRadii);
            DA.SetDataTree(3, treeCenters);
            DA.SetDataTree(4, treeDims);
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("B5D142EF-0158-45F4-A0C3-8B39B2A7EAC4"); }
        }

        protected override System.Drawing.Bitmap Icon
        {
            get { return IconLoader.Load("SegDisp.png"); }
        }
    }
}
